internal/services/v1/permissions.go

```diff
@@ -535,86 +536,157 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
 	}
 	usagemetrics.SetInContext(ctx, respMetadata)
 
-	stream := dispatchpkg.NewHandlingDispatchStream(ctx, func(result *dispatch.DispatchLookupSubjectsResponse) error {
-		foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
-		if !ok {
-			return fmt.Errorf("missing resource ID in returned LS")
+	var currentCursor *dispatch.Cursor
+	remainingConcreteLimit := 0
+
+	lsRequestHash, err := computeLSRequestHash(req)
+	if err != nil {
+		return ps.rewriteError(ctx, err)
+	}
+
+	if req.OptionalCursor != nil {
+		decodedCursor, err := cursor.DecodeToDispatchCursor(req.OptionalCursor, lsRequestHash)
+		if err != nil {
+			return ps.rewriteError(ctx, err)
 		}
+		currentCursor = decodedCursor
+	}
+
+	if req.OptionalConcreteLimit > 0 {
+		remainingConcreteLimit = int(req.OptionalConcreteLimit)
+	}
+
+	internalResponseCursor := &impl.DecodedCursor{
+		VersionOneof: &impl.DecodedCursor_V1{
+			V1: &impl.V1Cursor{
+				Revision:              atRevision.String(),
+				CallAndParametersHash: lsRequestHash,
+			},
+		},
+	}
+
+	for {
+		countSubjectsFound := 0
 
-		for _, foundSubject := range foundSubjects.FoundSubjects {
-			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream := dispatchpkg.NewHandlingDispatchStream(ctxWithCancel, func(result *dispatch.DispatchLookupSubjectsResponse) error {
+			foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+			if !ok {
+				return fmt.Errorf("missing resource ID in returned LS")
 			}
 
-			excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+			for _, foundSubject := range foundSubjects.FoundSubjects {
+				// Skip wildcards if requested they be skipped.
+				if req.WildcardOption == v1.LookupSubjectsRequest_WILDCARD_OPTION_EXCLUDE_WILDCARDS && foundSubject.SubjectId == tuple.PublicWildcard {
+					continue
+				}
+
+				excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
+				excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
+				for _, excludedSubject := range foundSubject.ExcludedSubjects {
+					excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+
+					resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
```

Comment:
not part of this PR, but for my own understanding: `FoundSubjects` can have itself `ExcludedFoundSubjects`, which can be caveated itself. It would look like to could be a composed caveat tree decorating the FoundSubjects. Shouldn't the caveats in the `FoundSubject` also be computed?

Reply:
Not sure I quite follow, but the non-deprecated fields on `FoundSubject` show it is caveated, as are the excluded ones (if necessary)

Reply:
What are the non-deprecated fields? I don't see any marked as such.

You can see below `ExcludedSubjects` is of type `FoundSubject`, which can have a `CaveatExpression`, which can happen recursively. If this is not possible because of the way it's built, why use the `FoundSubject` type in the first place?

```go
type FoundSubject struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	SubjectId        string               `protobuf:"bytes,1,opt,name=subject_id,json=subjectId,proto3" json:"subject_id,omitempty"`
	CaveatExpression *v1.CaveatExpression `protobuf:"bytes,2,opt,name=caveat_expression,json=caveatExpression,proto3" json:"caveat_expression,omitempty"`
	ExcludedSubjects []*FoundSubject      `protobuf:"bytes,3,rep,name=excluded_subjects,json=excludedSubjects,proto3" json:"excluded_subjects,omitempty"`
}
```

Reply:
All of these fields are deprecated:

```
// subject_object_id is the Object ID of the subject found. May be a `*` if
--
// a wildcard was found.
// deprecated: use `subject`
string subject_object_id = 2 [deprecated = true];
// excluded_subject_ids are the Object IDs of the subjects excluded. This list
// will only contain object IDs if `subject_object_id` is a wildcard (`*`) and
// will only be populated if exclusions exist from the wildcard.
// deprecated: use `excluded_subjects`
repeated string excluded_subject_ids = 3 [deprecated = true];
// permissionship indicates whether the response was partially evaluated or not
// deprecated: use `subject.permissionship`
LookupPermissionship permissionship = 4 [deprecated = true[(validate.rules) = {enum: {defined_only:true,not_in:[0]}}];];
// partial_caveat_info holds information of a partially-evaluated caveated response
// deprecated: use `subject.partial_caveat_info`
PartialCaveatInfo partial_caveat_info = 5 [deprecated = true[(validate.rules) = {message: {required:false}}];];

// subject_object_id is the Object ID of the subject found. May be a `*` if
	// a wildcard was found.
	// deprecated: use `subject`
	string subject_object_id = 2 [deprecated = true];
	// excluded_subject_ids are the Object IDs of the subjects excluded. This list
	// will only contain object IDs if `subject_object_id` is a wildcard (`*`) and
	// will only be populated if exclusions exist from the wildcard.
	// deprecated: use `excluded_subjects`
	repeated string excluded_subject_ids = 3 [deprecated = true];
	// permissionship indicates whether the response was partially evaluated or not
	// deprecated: use `subject.permissionship`
	[LookupPermissionship](https://buf.build/authzed/api/docs/main:authzed.api.v1#authzed.api.v1.LookupPermissionship) permissionship = 4 [deprecated = true[(validate.rules) = {
enum: {
defined_only:
true
,
not_in:
[0]
}
}];];
	// partial_caveat_info holds information of a partially-evaluated caveated response
	// deprecated: use `subject.partial_caveat_info`
	[PartialCaveatInfo](https://buf.build/authzed/api/docs/main:authzed.api.v1#authzed.api.v1.PartialCaveatInfo) partial_caveat_info = 5 [deprecated = true[(validate.rules) = {
message: {
required:
false
}
}];];
````

Reply:
Discussed Async: there should only be one level of `ExcludedSubjects`. Joey mentioned he didn't want to create yet another type just to denote there is only one level.